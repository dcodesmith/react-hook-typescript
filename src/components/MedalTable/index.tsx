import { getCode } from 'country-list';
import React from 'react';

interface IMedals {
  gold: string;
  silver: string;
  bronze: string;
}

interface IMedalData {
  country: string;
  medals: IMedals;
}

interface ITableProps {
  data: IMedalData[];
  onDelete(country: string): void;
  onSelect?(country: string): void;
}

const MedalTable = ({ data, onDelete, onSelect }: ITableProps) => {
  const getCountryCode = (country: string) => {
    const countryCode = getCode(country);

    if (countryCode) {
      return countryCode.toLocaleLowerCase();
    }

    return '';
  };

  return (
    <div className="grid">
      <table className="medal-table">
        <thead>
          <tr>
            <th className="medal-table__header__cell">#</th>
            <th className="medal-table__header__cell">Country</th>
            <th className="medal-table__header__cell">
              <span className="medal-table__header__cell--gold" />
            </th>
            <th className="medal-table__header__cell">
              <span className="medal-table__header__cell--silver" />
            </th>
            <th className="medal-table__header__cell">
              <span className="medal-table__header__cell--bronze" />
            </th>
            <th className="medal-table__header__cell">Total</th>
            <th className="medal-table__header__cell">&nbsp;</th>
          </tr>
        </thead>
        <tbody className="medal-table__body">
          {data.map(({ country, medals: { gold, silver, bronze } }, index) => (
            <tr
              className="medal-table__row"
              key={country}
              onClick={() => onSelect && onSelect(country)}
            >
              <td className="medal-table__body__cell">{index + 1}</td>
              <td className="medal-table__body__cell">
                <span
                  className={`flag-icon flag-icon-${getCountryCode(country)}`}
                />
                <span className="medal-table__body__country">{country}</span>
              </td>
              <td className="medal-table__body__cell">{gold}</td>
              <td className="medal-table__body__cell">{silver}</td>
              <td className="medal-table__body__cell">{bronze}</td>
              <td className="medal-table__body__cell">
                {Number(gold) + Number(silver) + Number(bronze)}
              </td>
              <td className="medal-table__body__cell">
                <div className="medal-table__body__cell__actions">
                  <button
                    className="medal-table__delete__button"
                    onClick={event => {
                      event.stopPropagation();
                      onDelete(country);
                    }}
                  >
                    Delete
                  </button>
                </div>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default MedalTable;
