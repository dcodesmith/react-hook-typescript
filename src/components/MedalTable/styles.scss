@mixin medal-style {
  display: inline-flex;
  width: 20px;
  height: 20px;
  border-radius: 50%;
}

.grid {
  margin: 0 10px;
}

.medal-table {
  width: 100%;
  border-collapse: collapse;

  &__row {
    &:hover {
      background: #efefef;
      cursor: pointer;
    }
  }

  &__body {
    text-align: center;

    &__cell {
      border-bottom: 1px solid #cfcfcf;
      padding: 5px 0;

      &:nth-child(2) {
        text-align: left;
      }

      &__actions {
        display: flex;
      }
    }

    &__country {
      display: inline-flex;
      margin-left: 10px;
    }
  }

  &__delete__button {
    color: red;
    border: 1px solid red;
    background: transparent;
  }

  &__edit__button {
    border: 1px solid #0989ef;
    background: #0989ef;
    margin-left: 5px;
  }

  &__header__cell {
    font-weight: 500;
    padding: 10px 0;
    border-bottom: 1px solid #cfcfcf;

    &--gold {
      @include medal-style();
      background: gold;
    }

    &--silver {
      @include medal-style();
      background: silver;
    }

    &--bronze {
      @include medal-style();
      background: brown;
    }

    &:nth-child(2) {
      text-align: left;
    }
  }
}

@media (min-width: 576px) {
  .grid {
    margin: 0;
  }
}
