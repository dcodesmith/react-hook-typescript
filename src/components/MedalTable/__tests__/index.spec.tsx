import { shallow } from 'enzyme'
import React from 'react'

import MedalTable from '..'

const onDeleteSpy = jest.fn()
const stopPropagationSpy = jest.fn()
const props = {
  data: [
    {
      country: 'China',
      medals: {
        gold: 100,
        silver: 0,
        bronze: 0,
      },
    },
  ],
  onDelete: onDeleteSpy,
}

describe('Given a MedalTable Component', () => {
  describe('When rendered', () => {
    let MedalTableComponent: any

    beforeAll(() => {
      MedalTableComponent = shallow(<MedalTable {...(props as any)} />)
    })

    it('should render a list of countries with their medals', () => {
      expect(MedalTableComponent).toMatchSnapshot()
    })

    describe('and a country is deleted', () => {
      beforeAll(() => {
        MedalTableComponent.find('button').simulate('click', {
          stopPropagation: stopPropagationSpy,
        })
      })

      it('should invoke the onDelete function with the countries name as an argument', () => {
        expect(stopPropagationSpy).toHaveBeenCalledTimes(1)
        expect(onDeleteSpy).toHaveBeenCalledWith('China')
      })
    })
  })
})
