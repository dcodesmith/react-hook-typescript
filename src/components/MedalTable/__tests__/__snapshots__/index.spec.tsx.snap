// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MedalTable Component When rendered should render a list of countries with their medals 1`] = `
<div
  className="grid"
>
  <table
    className="medal-table"
  >
    <thead>
      <tr>
        <th
          className="medal-table__header__cell"
        >
          #
        </th>
        <th
          className="medal-table__header__cell"
        >
          Country
        </th>
        <th
          className="medal-table__header__cell"
        >
          <span
            className="medal-table__header__cell--gold"
          />
        </th>
        <th
          className="medal-table__header__cell"
        >
          <span
            className="medal-table__header__cell--silver"
          />
        </th>
        <th
          className="medal-table__header__cell"
        >
          <span
            className="medal-table__header__cell--bronze"
          />
        </th>
        <th
          className="medal-table__header__cell"
        >
          Total
        </th>
        <th
          className="medal-table__header__cell"
        >
          Â 
        </th>
      </tr>
    </thead>
    <tbody
      className="medal-table__body"
    >
      <tr
        className="medal-table__row"
        key="China"
        onClick={[Function]}
      >
        <td
          className="medal-table__body__cell"
        >
          1
        </td>
        <td
          className="medal-table__body__cell"
        >
          <span
            className="flag-icon flag-icon-cn"
          />
          <span
            className="medal-table__body__country"
          >
            China
          </span>
        </td>
        <td
          className="medal-table__body__cell"
        >
          100
        </td>
        <td
          className="medal-table__body__cell"
        >
          0
        </td>
        <td
          className="medal-table__body__cell"
        >
          0
        </td>
        <td
          className="medal-table__body__cell"
        >
          100
        </td>
        <td
          className="medal-table__body__cell"
        >
          <div
            className="medal-table__body__cell__actions"
          >
            <button
              className="medal-table__delete__button"
              onClick={[Function]}
            >
              Delete
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
